src/staker.cairo

```diff
@@ -333,5 +355,66 @@ pub mod Staker {
             let now = get_block_timestamp();
             self.get_average_delegated(delegate, now - period, now)
         }
+
+        fn get_cumulative_seconds_per_total_staked_at(self: @ContractState, timestamp: u64) -> u256 {
+            if let Option::Some((log_record, idx)) = self.staking_log.find_in_change_log(timestamp) {
+                let total_staked = if (idx == self.staking_log.len() - 1) {
+                    // if last rescord found
+                    self.total_staked.read()
+                } else {
+                    // otherwise calculate using cumulative_seconds_per_total_staked difference
+                    let next_log_record = self.staking_log.at(idx+1).read();
+                    
+                    // substract fixed point values
+                    let divisor = next_log_record.cumulative_seconds_per_total_staked - log_record.cumulative_seconds_per_total_staked;
+                    assert(divisor.high < MAX_FP, 'FP_OVERFLOW');
+                    
+                    if divisor.is_zero() {
+                        return 0_u64.into();
+                    }
+
+                    let diff_seconds: u128 = ((next_log_record.timestamp - log_record.timestamp) / 1000).into();
+
+                    // Divide u64 by fixed point
+                    let (total_staked_fp_medium, _) = u512_safe_div_rem_by_u256(
+                        u512 { limb0: 0, limb1: 0, limb2: 0, limb3: diff_seconds },
+                        divisor.try_into().unwrap()
+                    );
+                        
+                    let total_staked_fp = u256 {
+                        low: total_staked_fp_medium.limb1,
+                        high: total_staked_fp_medium.limb2,
+                    };
+
+                    assert(total_staked_fp.high < MAX_FP, 'FP_OVERFLOW');
+
+                    // round value
+                    total_staked_fp.high + if (total_staked_fp.low >= HALF) {
+                        1
+                    } else {
+                        0
+                    }
+                };
+                
+                let seconds_diff = (timestamp - log_record.timestamp) / 1000;
```

Comment:
block timestamps are in seconds arent they? why are we dividing by 1k

Reply:
This is absolutely right. I can't believe tests did not catch it.